pkg/ingester/ingester_v2_backfill.go

```diff
@@ -0,0 +1,499 @@
+	b.dbs[idx] = nil
+	b.mtx.Unlock()
+
+	defer func() {
+		if err == nil {
+			if cerr := db.db.Close(); cerr != nil {
+				err = cerr
+			}
+		}
+
+		b.mtx.Lock()
+		if err != nil {
+			b.dbs[idx] = db
+		} else {
+			if b.dbs[0] == nil && b.dbs[1] == nil {
+				b.metrics.numBackfillTSDBsPerUser.DeleteLabelValues(b.userID)
+				b.metrics.numUsersWithBackfillTSDBs.Dec()
+			} else {
+				b.metrics.numBackfillTSDBsPerUser.WithLabelValues(b.userID).Dec()
+			}
+		}
+		b.mtx.Unlock()
+	}()
+
+	h := db.db.Head()
+	if err := db.db.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), h.MaxTime())); err != nil {
+		return errors.Wrapf(err, "compact backfill TSDB, dir:%s", db.db.Dir())
+	}
+
+	if db.db.shipper != nil {
+		uploaded, err := db.db.shipper.Sync(context.Background())
+		if err != nil {
+			return errors.Wrapf(err, "ship backfill TSDB, uploaded:%d, dir:%s", uploaded, db.db.Dir())
+		}
+		level.Debug(util.Logger).Log("msg", "shipper successfully synchronized backfill TSDB blocks with storage", "user", db.db.userID, "uploaded", uploaded, "backfill_dir", db.db.Dir())
+	}
+
+	return nil
+}
+
+func (b *backfillTSDB) isIdle(timeout time.Duration) bool {
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	idle := false
+	for i := 0; i < 2; i++ {
+		if b.dbs[i] != nil {
+			idle = idle || b.dbs[i].db.isIdle(mtime.Now(), timeout)
+		}
+	}
+	return idle
+}
+
+func (b *backfillTSDB) compactAndShipIdleTSDBs(timeout time.Duration) error {
+	b.compactMtx.Lock()
+	defer b.compactMtx.Unlock()
+
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	var merr tsdb_errors.MultiError
+	for i := 0; i < 2; i++ {
+		b.mtx.RLock()
+		if b.dbs[i] == nil || !b.dbs[i].db.isIdle(mtime.Now(), timeout) {
+			b.mtx.RUnlock()
+			continue
+		}
+		db := b.dbs[i]
+		b.mtx.RUnlock()
+
+		err := b.compactAndShipAndCloseDB(i)
+		if err != nil {
+			merr.Add(err)
+			continue
+		}
+
+		if err := os.RemoveAll(db.db.Dir()); err != nil {
+			// TODO(codesome): Add a metric for this to alert on.
+			level.Error(util.Logger).Log("msg", "failed to delete backfill TSDB dir in compactAndShipIdleTSDBs", "user", db.db.userID, "dir", db.db.Dir())
+		}
+	}
+
+	return merr.Err()
+}
+
+func (a *backfillAppender) timeRangesForTimestamp(ts int64) (int64, int64) {
+	step := a.backfillTSDB.backfillAge.Milliseconds()
+	start := (ts / step) * step
+	end := start + step
+	return start, end
+}
+
+// getBucketName returns the string representation of the bucket.
+// YYYY_MM_DD_HH_YYYY_MM_DD_HH
+func getTSDBName(start, end int64) string {
+	startTime := model.Time(start).Time().UTC()
+	endTime := model.Time(end).Time().UTC()
+
+	return fmt.Sprintf(
+		"%04d_%02d_%02d_%02d_%04d_%02d_%02d_%02d",
+		startTime.Year(), startTime.Month(), startTime.Day(), startTime.Hour(),
+		endTime.Year(), endTime.Month(), endTime.Day(), endTime.Hour(),
+	)
+}
+
+func overlapsOpenInterval(mint1, maxt1, mint2, maxt2 int64) bool {
+	return mint1 < maxt2 && mint2 < maxt1
+}
+
+func getBackfillTSDBRanges(tsdbName string) (int64, int64, error) {
```

Comment:
You could simplify this a lot (and make it even stronger) using a single regular expression to extract all data.

Reply:
Thanks for the regex idea. I was facing some issues with time.Parse. Regex should be simpler.